### DISCUSSION TITLE
Can I implement a file clipboard from a sample client?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8283] [Q&A]
Hello.

I am making RDP function available on the web using sample client and JNI.

Based on the sample client, the use of JNI in the Android version is implemented as an example (the implementation is Java, not Android)

The implementation of text clipboard synchronization between servers and clients is complete.

In the web server (browser), the user's current clipboard is delivered to the server using a web socket, and the JNI module that I implemented is used to synchronize the clipboard. 

However, the problem now is file clipboard synchronization.

I know that the Android client and the Mac client does not have a file clipboard implementation.

Since the connection between the client and the server is based on the web browser, I think it can be implemented without being affected by the operating system if it is possible to implement the file clipboard using JNI..

Q - If file data can be delivered like text data using JNI , can file clipboard synchronization be implemented on the sample client?

Q - And If implementation is possible, could you give me a simple advice on the difference between implementing a text clipboard and a file clipboard?

I am having difficulty analyzing the code to implement the file clipboard, and I want to shorten the time.

Thank you for reading my question.

**Comment by akallabeth** [comment]
Not too familiar with websockets and limitations of clipboard in browsers, but for the FreeRDP part:
* The clipboard is asynchronous, e.g. first the formats / file lists are announced (either from server -> client or client -> server) and then (upon request, e.g. paste) the data is transferred.
* There is a huge difference between client -> server and server -> client clipboard. 
   * client -> server clipboard  just needs to send the data when the server requests it. (if you have e.g. a file, just read and send starting at the requested offset)
   * server -> client is more complex as you also need to present the file data to the local os / browser / ... `xfreerdp` uses fuse for that, so no idea what you need. The only other way is copy everything to local first and then present it (really laggy, not really usable, needs a temporary file of same size, problem with iso images or similar if the disk is not big enough)

**Reply by kimcoldwater** [reply]
I use JS on the web server.

Using JS clipboard API, you can overwrite the local clipboard of the client (web user) directly as long as you have file data coming from the server -> client.

This means that you only need file data, but you don't need a file path.

I'll try more.

And do I have to implement `cliprdr_server_file_content_request` And `response` to implement file clipboard?

Android and Mac have empty functions.

**Reply by akallabeth** [reply]
1. ok, you need the `full file data` or just whatever is requested?
1. You are talkaing about `server -> client` file transfer? then the client sends a request (see `xf_cliprdr_send_client_file_contents`) and the server answers it with a response (see `xf_cliprdr_server_file_contents_response`)
2. if you are talking about `client -> server` file data transfer then you need that function (see `xf_cliprdr_server_file_contents_request`) to return the actual file data (or size)

**Reply by kimcoldwater** [reply]
I am implementing two-way clipboard synchronization between server and client.

I analyzed the code and conducted the test according to your kind advice for the past few days.

I understood that the key to file conversion is for `FileGroupDescriptorW` and I looked up a lot of data.

I also checked the `xf_clipdr_server_file_content_response`.

First of all, I'm focusing on the `server-> client`.

 Copying a file does not callback `ServerFileContentsResponse`.

only callback  -> `ServerFormatDataResponse`, but after logging, I don't think I'm getting the right file data here.

I don't need Fuse or anything but file data and file length. I will send it to java using jni and paste it directly on the user's clipboard. (Route required x)

Q1. How do I get ServerFileContentsResponse called callback?

Of course, `init` correctly specifies the function.

Q2. Is BYTE* data in FileDescriptor format?

I wonder what type of data comes from the server side to the client. 

This is an important part when data is transferred to c->java.

**Reply by akallabeth** [reply]
1. First you get the file format announcement (that is indeed a `FileGroupDescriptorW`)
2. from that you extract the file information (and `offset`, look at `cliprdr_serialize_file_list_ex`)
3. Now you can request specific file data (e.g. size and the file itself)
4. Also note that you need the `LockClipboardData` to tell the server to keep the file list until you have copied the data you are interested in. (otherwise the next copy operation will remove the file from the server clipboard and your file data requests will fail)

**Reply by kimcoldwater** [reply]
```
	formatId = ClipboardRegisterFormat(tfc->clipboard, "FileGroupDescriptorW");
			data = (void *)ClipboardGetData(tfc->clipboard, formatId, &size);
			FILEDESCRIPTORW* file_array = (FILEDESCRIPTORW*)data;
                		UINT32 file_count = size / sizeof(FILEDESCRIPTORW);

                		data = NULL;
                        	size = 0;
                         cliprdr_serialize_file_list_ex(CB_STREAM_FILECLIP_ENABLED , file_array,
                		file_count, &data, &size);

                JNIEnv* env;
                jni_attach_thread(&env);
	       jbyteArray array = (*env)->NewByteArray(env, sizeof(unsigned char) * size);
		(*env)->SetByteArrayRegion(env, array, 0, sizeof(unsigned char) * size , data);
                freerdp_callback("jni_callback_clipboard_file", "(J[B)V",instance, array);
                (*env)->DeleteLocalRef(env, array);
                free(data);
```

Based on your advice, I tried adding the logic to `clipdr_server_format_data_response` and tried many things.

I've also added a lot of logs to see what formatId I get when I copy files from the server.

However, the length of `formatDataResponse->dataLen`, which was copying several files, is fixed to 4, and an undefined formatId is entered.

CF_OEMTEXT is normally caught as formatId when copying text, but no Id defined as CF_* is captured when copying files.

I made it according to the clipboard channel configuration of the current Android implementation, but can't I get the file clipboard data in this way?

If you look at the implementation of the x11 or Windows clipboard, there are a few more functions on Android, but is that necessary?

And I'm implementing it in mac.
I'm implementing a sample client, not a mac client, but can this be a problem?

I don't want to give up the file clipboard functionality, but it's time consuming and difficult.

so sad.. :(

**Reply by akallabeth** [reply]
No.
You do correct with registering the format (and remembering the format ID, but you can only get data that was announced from the server.
Best check the behavior of `xfreerdp` with fuse first to see how it is meant to work.

**Reply by kimcoldwater** [reply]
Clipboard with mouse returns an undefined format.

While searching, I found a mailing list recommending clipboard with control + c.

However, if you copy the clipboard like this` [com].freerdp.channels.clipdr.client] - ServerFormatList: numFormats: 0` is output and clipboard data cannot be imported.

Why does numFormats output 0?

I was mistaken that most of the file clipboard came in `FILEDESCRIPTORW` format.

In fact, when I looked at format->formatName as log in response, a completely different format was being output. (Undefined)

You can use it with registerFormat, but... Looking at the length of the data, I don't think it's data for the file.

Can you give me some advice on what I'm missing?

**Reply by akallabeth** [reply]
you interpret the metadata for file transfer as file data?

Also, never seen any difference between `<ctrl>+c` to mouse copy, so there is something else off...

There is the normal clipboard which contains formats like text, bitmap, ...
This normal clipboard will have the `FileDescriptorW` list if you copied files.
With this list you can start transferring file data with the method previously mentioned, e.g. client sends ContentsRequest, server answers with `ContentsResponse`

**Reply by kimcoldwater** [reply]
Can I remove Atom or X function from the clipboard part of x11?

Will it still work if I implement it this way?

I have confirmed that the X11 works in the format you mentioned, but is it possible to implement it without configuring the Xserver?

I don't want to create an X11 server while implementing a headless client. Display variables cannot be included.

And the wayland version seems more convenient to refer to, is the file clipboard implemented and working?

I tried to implement the clipboard by referring to the wayland version, but I couldn't get the file format normally.

The text worked fine, but when I copied the file, I was only able to get the format shown in the issue below.

https://github.com/FreeRDP/FreeRDP/issues/2441

And so did the Android implementation.

However, if you run the x11 version , the FileDescriptorW format is imported normally.

I feel sorry that I am having a hard time implementing your kind advice.

I refer to the master version of X11 with fuse, and tried to implement it similarly, 

but I'm having a hard time removing the X_function.
